File: server/apps/sync-service/src/main/java/com/sharetimer/syncservice/application/port/in/TimerUseCase.java
package com.sharetimer.syncservice.application.port.in;

import java.time.Instant;
import org.springframework.http.codec.ServerSentEvent;
import reactor.core.publisher.Flux;

public interface TimerUseCase {

  Flux<ServerSentEvent<Object>> subscribe(String timerId);

  void processTimerExpiration(String timerId);

  void updateTargetTime(String timerId, Instant updatedAt, Instant serverTime,
      Instant newTargetTime);

  void addTimestamp(String timerId, Instant targetTime, Instant capturedAt);

}

File: server/apps/sync-service/src/main/kotlin/com/sharetimer/syncservice/application/port/input/TimerUseCase.kt
package com.sharetimer.syncservice.application.port.input

import kotlinx.coroutines.flow.Flow
import org.springframework.http.codec.ServerSentEvent
import java.time.Instant

interface TimerUseCase {
    fun subscribe(timerId: String): Flow<ServerSentEvent<Any>>

    fun processTimerExpiration(timerId: String)

    fun updateTargetTime(
        timerId: String,
        updatedAt: Instant,
        serverTime: Instant,
        newTargetTime: Instant,
    )

    fun addTimestamp(
        timerId: String,
        targetTime: Instant,
        capturedAt: Instant,
    )
}
